### ISSUE TITLE
0dyseus@SysmonitorByOrcus: Installing applets not compatible with vertical panel froze the session

### ISSUE THREAD

**Issue by DiWa51** [linuxmint/cinnamon-spices-applets#24]
While trying to find applets compatible with vertical panel, I've faced the following annoying issue:

1. After installing multicore-sys-monitor@ccadeptic23, then netusagemonitor@pdcurtis, I installed a 3rd one (0dyseus@SysmonitorByOrcus)
2. For each of them I had a popup stating the applet is not compatible with vertical panel.
3. For the last applet though, the **screen froze** on the popup and I couldn't do anything but moving the mouse's cursor (no mouse click, no keyboard). 
4. I had to hard reboot, but each time I logged in, everything but the mouse was instantly freezing (after the welcome sound). After some time my fan was blowing meaning the CPU was probably running fast.
5. Finally to sort this out, I had to delete the .local/share/cinnamon/applets folder from a text session.

If there are logs that could be helpful tell me how to find them and I'll provide them.

Cheers

**Comment by jaszhix** [comment]
You can look for errors in both ```~/.cinnamon/glass.log``` and ```~/.xsession-errors.log``` files. You may need to add ```LimitSessionOutput=false``` under ```[debug]``` in ```/etc/mdm/mdm.conf``` to bypass the log size limit in ```.xsession-errors```.

**Comment by DiWa51** [comment]
These logs just report things after I fixed my pb, i.e. after a reboot. 

Can they survive a reboot? Is it the purpose of `LimitSessionOutput=false`

Given my only solution to reproduce the issue is to log in the UI Session and the only way I found to get out of the freeze is to hard reboot this might be useless to try to reproduce.

**Comment by jaszhix** [comment]
They should stay persistent after reboots.

@pdcurtis @Odyseus

**Comment by collinss** [comment]
Actually, neither log will survive a reboot. .xsession-errors is emptied when a new x-session is started, and glass.log is emptied when Cinnamon is restarted.

@DiWa51 Are you able to switch to a TTY? If so, you could make a copy of .xsession-errors with a different name from command line. This would allow it to survive a reboot, though even then, there is a good chance that there wont be anything there. Freezes often don't leave any errors, as they are often caused by valid code with faulty logic. (The increased cpu usage suggests that might indeed be the case here). They can be particularly difficult to find and fix. If you could find an applet that reliably reproduces the issue for you, that would probably help a lot more.

**Comment by jaszhix** [comment]
@collinss Ah, my mistake. It seems at least for me my .xsession-errors is not clearing after reboots. I had to increase the log limit to see new errors after a while.

**Comment by Odyseus** [comment]
Hello, everybody.

@jaszhix & @collinss: I can reproduce what @DiWa51 is reporting with my applet ([0dyseus@SysmonitorByOrcus](https://github.com/linuxmint/cinnamon-spices-applets/tree/master/0dyseus%40SysmonitorByOrcus/files/0dyseus%40SysmonitorByOrcus)). If I try to place the applet into a vertical panel, I get the normal popup informing me that the applet isn't suitable for vertical panels and the system freeze.

This is the content of **.xsession-errors** at the exact time of the freeze.

```
(nemo:4196): GVFS-RemoteVolumeMonitor-WARNING **: Owner of volume monitor org.gtk.vfs.UDisks2VolumeMonitor disconnected from the bus; removing drives/volumes/mounts

(nemo:4196): GVFS-RemoteVolumeMonitor-WARNING **: Owner of volume monitor org.gtk.vfs.AfcVolumeMonitor disconnected from the bus; removing drives/volumes/mounts

(nemo:4196): GVFS-RemoteVolumeMonitor-WARNING **: Owner of volume monitor org.gtk.vfs.GoaVolumeMonitor disconnected from the bus; removing drives/volumes/mounts

(nemo:4196): GVFS-RemoteVolumeMonitor-WARNING **: Owner of volume monitor org.gtk.vfs.MTPVolumeMonitor disconnected from the bus; removing drives/volumes/mounts

(nemo:4196): GVFS-RemoteVolumeMonitor-WARNING **: Owner of volume monitor org.gtk.vfs.GPhoto2VolumeMonitor disconnected from the bus; removing drives/volumes/mounts

** (cinnamon-settings-daemon:4101): WARNING **: Name taken or bus went away - shutting down

** (cinnamon-settings-daemon:4101): WARNING **: Name taken or bus went away - shutting down
Terminated
```

Placing the applet into an horizontal panel works normally **but**, when I try to remove it from the horizontal panel, the system freeze with a similar error logged into **.xsession-errors**.

```
cinnamon-session[4662]: Gdk-WARNING: t+478,68021s: cinnamon-session: Fatal IO error 11 (Resource not available temporarily) on X server :1.


** (cinnamon-settings-daemon:4700): WARNING **: Name taken or bus went away - shutting down

** (cinnamon-settings-daemon:4700): WARNING **: Name taken or bus went away - shutting down

(cinnamon-settings-daemon:4700): Gdk-WARNING **: cinnamon-settings-daemon: Fatal IO error 11 (Resource not available temporarily) on X server :1.
```

All this is very strange, because removing the applet from the panel on Cinnamon 2.8 (Linux Mint 17.3) or Cinnamon 3.0.7 (Linux Mint 18) works perfectly without any freeze nor errors/warnings logged anywhere. And when I switched this applet to *multiversion* a couple of months ago, the applet worked fine also on Cinnamon 3.2.x.

Just in case it's useful, my Cinnamon 3.2.x system is in a VirtualBox machine. It used to be a LM18 system with [this PPA](https://github.com/linuxmint/Cinnamon/wiki/Trying-out-the-latest-Cinnamon-without-building-it.) and then I updated it to LM18.1 through the updates manager (still using the same PPA and up-to-date).

@DiWa51

You don't need to completely reboot your system nor to completely remove your applets folder. You just have to go to a TTY (<kbd>Ctrl</kbd> + <kbd>Alt</kbd> + <kbd>F1</kbd>), log in and then rename the **~/.xsession-errors** and **~/.cinnamon/glass.log** files (so they aren't overwritten on next Cinnamon start) and then rename the applet's folder (so the system would not freeze on next Cinnamon). When you finish, just use the `startx` command to start Cinnamon. Remember to remove the applet from the panel after starting Cinnamon and renaming back the applet's folder.

**Comment by DiWa51** [comment]
This is good news in a sense!

I knew there was a shortcut to actually switch to text mode but as I didn't know it by heart I couldn't search internet on my frozen computer, _from this you can guess I'm not in a VM_. 
Just crazy how dependent we are on these machines ;)

Thanks for the info I'll try to reproduce it and grab the logs, but not in the following days unfortunately...

**Comment by Odyseus** [comment]
Hello, everybody.

@DiWa51: I fixed the system freeze on my applet (0dyseus@SysmonitorByOrcus). A pulll request is queued to be approved. It seems that the error messages that I posted in my previous message weren't related at all with the system freeze produced by my applet. Because the next day, when doing tests, there were no errors at all when the system froze.

Although the system does not freeze when placing this applet on a vertical panel, it is still not compatible for panels of said orientation. And I don't know if it will ever be.

**P.S.:** It seems that the freeze was caused by the usage of the **Mainloop.timeout_add()** function without storing it and without *cleaning it* at applet removal.
